### PR TITLE
Add door electronics access configuration menu

### PR DESCRIPTION
**Pull request by c4llv07e**
## About the PR

Adds a door electronics menu that allows you to configure access to it. The configuration is saved even after disassembly (video below).

This is the third PR trying to add programmable access for door electronics (#14680 and #12424), so it's more of an idea than a final result.

Unlike #14680, no access is required to change it. I don't know if this is really required, because I can't imagine how it could be abused.

I tried my best to check this code against the required standards, but I still might be missing something.
There are some pieces of code from other things in the game (such as the ID Card Console, Gravity Generator UI, Forensics Scanner, etc.), because of this, some solutions may not be as good as they could be.

**Media**

https://github.com/space-wizards/space-station-14/assets/38111072/503e4a0b-9f8f-4662-be54-5a64fdfba880

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl: kseandi
- add: NT has declassified the documentation for door electronics, now anyone can configure its access using network configurator or multitool.

### CODE REVIEW

**Review by 0x6273 — CHANGES_REQUESTED** [review]
Use file-scoped namespaces for new files.

**Comment by c4llv07e** [comment]
Edited

**Comment by Cheackraze** [comment]
So far this seems like the most functional attempt, if a bit simplistic.  It avoids many of the logic traps the other PRs introduced.  While it may not make much sense to just be able to pull up a UI off a bare circuit board, this seems like the most functional way that won't have unintended consequences like allowing maints access on every door etc.

Maybe if you needed to use a network configuration tool on the board to open the UI?  Not necessary buy it will make the 'immersion' crowd happy

**Comment by c4llv07e** [comment]
I think it's a cool idea because it will add a new use case to the network configuration tool, which is great. I'll try to do it.

**Comment by c4llv07e** [comment]
Now it should work.
I tested it and it's working better than I expected. This codebase is amazing! I really don't understand why UI API is called "wasteland" in the docs.

**Comment by dmnct** [comment]
When you first pop out the electronics of a mapped door it "loses" it's mapped configuration

**Comment by lzk228** [comment]
how it would looks for mapping?

**Comment by c4llv07e** [comment]
@dmnct As far as I know, you cannot determine if an entity is inside a container in another entity, so this change requires editing the DoorSystem code. So it may be not as clean as it should be.

**Comment by c4llv07e** [comment]
@lzk228 Huh, here come some problems:
At the moment, there are child prototypes of the Airlock entity that have an additional AccessReaderComponent with different accesses. If we want to move the AccessReaderComponent from the door to the electronics, we also need to change these prototypes. This will be more difficult, because now we don't need to add a component, but edit it.
Right now I can't find any good solution that doesn't break the inheritance hierarchy.

**Comment by lzk228** [comment]
i think that would be cool to change airlock access in admin-trick menu

**Comment by c4llv07e** [comment]
> i think that would be cool to change airlock access in admin-tricks menu

Maybe, but I can't imagine how it can be done, and it seems to me that it should be done in another PR (Read it as "This is my first PR, and I'm too scared to make it bigger" (: ).

File: Content.Shared/Doors/Electronics/SharedDoorElectronicsSystem.cs
Comment (AJCM-git) [inline Content.Shared/Doors/Electronics/SharedDoorElectronicsSystem.cs:1]: This is not used anywhere

File: Content.Shared/Doors/Electronics/SharedDoorElectronicsComponent.cs
Comment (AJCM-git) [inline Content.Shared/Doors/Electronics/SharedDoorElectronicsComponent.cs:1]: neither is this (components should only contain fields, these messages, ui keys and stuff shouldnt be inside a class, put em in the same namespace (not class, namespace, AKA just in the same file) as the DoorElectronicsComponent)

File: Content.Server/Doors/Systems/DoorSystem.cs [hunk 1]
```diff
@@ -280,5 +282,22 @@ protected override void CheckDoorBump(DoorComponent component, PhysicsComponent
...
+            return;
+
+        foreach (var entity in boardContainer.ContainedEntities)
+        {
+            if (TryComp<AccessReaderComponent>(entity, out var boardAccess))
+                boardAccess.AccessLists = access.AccessLists;
+        }
+
+        EntityManager.RemoveComponent<AccessReaderComponent>(uid);
+    }
```
Comment (AJCM-git) [inline Content.Server/Doors/Systems/DoorSystem.cs:302]: this is a workaround to just tweaking the airlock prototypes to remove their innate access and put door boards replacements

File: Content.Client/Doors/Electronics/DoorElectronicsConfigurationMenu.xaml.cs
Comment (AJCM-git) [inline Content.Client/Doors/Electronics/DoorElectronicsConfigurationMenu.xaml.cs:1]: This is a copypaste from IDCardConsoleWindow, just make a new configurable AccessLevelControl or something and use it for both

**Review by AJCM-git — CHANGES_REQUESTED** [review]

File: Content.Client/Doors/Electronics/DoorElectronicsConfigurationMenu.xaml.cs
Comment (c4llv07e) [reply Content.Client/Doors/Electronics/DoorElectronicsConfigurationMenu.xaml.cs:1]: Done, needs review

File: Content.Server/Doors/Systems/DoorSystem.cs (hunk 1 above)
Comment (c4llv07e) [reply Content.Server/Doors/Systems/DoorSystem.cs:302]: Yes, as I talked here https://github.com/space-wizards/space-station-14/pull/17778#issuecomment-1616680258 , I can't find any good solution how to move AccessReaderComponent from airlock prototypes to their door electronics.

File: Resources/Prototypes/Entities/Objects/Devices/Electronics/door.yml [hunk 2]
```diff
@@ -12,5 +12,10 @@
       tags:
         - DoorElectronics
```
[the remark this answers is not in the document]
Comment (c4llv07e) [reply Resources/Prototypes/Entities/Objects/Devices/Electronics/door.yml]: Done

File: Content.Shared/Doors/Electronics/SharedDoorElectronicsSystem.cs
Comment (c4llv07e) [reply Content.Shared/Doors/Electronics/SharedDoorElectronicsSystem.cs:1]: Used once in `Content.Client/Doors/Electronics/DoorElectronicsBoundUserInterface.cs`. Is it good to keep separate Sawmill just for one log?

**Comment by c4llv07e** [comment]
Hmm, I think I've done everything. There's still an `AccessLevelControl` that should be reviewed, but I think everything else is fine now.

File: Content.Server/Doors/Electronics/Systems/DoorElectronicsSystem.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+        if (!TryComp<NetworkConfiguratorComponent>(args.Used, out var networkConfigurator))
+            return;
+
+        if (!TryComp(args.User, out ActorComponent? actor))
+            return;
+
+        args.Handled = true;
+
+        _uiSystem.TryOpen(uid, DoorElectronicsConfigurationUiKey.Key, actor.PlayerSession);
```
Comment (AJCM-git) [inline Content.Server/Doors/Electronics/Systems/DoorElectronicsSystem.cs:82]: This review is not a requirement but a recommendation, so you decide whether or not you want to do this. Replace this with an `ActivatableUIComponent`, tho it needs to support interacting *with* something to open the UI, this should be easy to do as i *think* you only need to add an `EntityWhitelist?` datafield to the component and then handle `InteractUsingEvent` on the `ActivatableUISystem` (filtering the used entity through `whitelistField.IsValid(args.Used, EntityManager)` method first, of course), and finally set the whitelist datafield to NetworkConfigurator in the prototype, feel free to ask any question.

File: Content.Server/Doors/Electronics/Systems/DoorElectronicsSystem.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        }
+
+        var state = new DoorElectronicsConfigurationState(accesses);
+
+        _uiSystem.TrySetUiState(uid, DoorElectronicsConfigurationUiKey.Key, state);
+    }
+
+    private void OnChangeConfiguration(EntityUid uid,
+                                       DoorElectronicsComponent component,
+                                       DoorElectronicsUpdateConfigurationMessage args)
```
Comment (AJCM-git) [inline Content.Server/Doors/Electronics/Systems/DoorElectronicsSystem.cs:46]: dont newline event handler params like this (there are other cases like this in the pr)

File: Content.Server/Doors/Electronics/Systems/DoorElectronicsSystem.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        SubscribeLocalEvent<DoorElectronicsComponent, DoorElectronicsUpdateConfigurationMessage>(OnChangeConfiguration);
+        SubscribeLocalEvent<DoorElectronicsComponent, DoorElectronicsRefreshUiMessage>(OnRefreshUi);
+        SubscribeLocalEvent<DoorElectronicsComponent, InteractUsingEvent>(OnInteractUsing);
+    }
+
+    public void UpdateUserInterface(EntityUid uid, DoorElectronicsComponent component)
+    {
+        List<string> accesses = new List<string>();
+
+        if (TryComp<AccessReaderComponent>(component.Owner, out var accessReader))
```
Comment (AJCM-git) [inline Content.Server/Doors/Electronics/Systems/DoorElectronicsSystem.cs:28]: .Owner is deprecated, pls if you copy code from somewhere else at least make sure there are no obsolete warnings

File: Content.Server/Doors/Electronics/Systems/DoorElectronicsSystem.cs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    {
+        base.Initialize();
+        SubscribeLocalEvent<DoorElectronicsComponent, DoorElectronicsUpdateConfigurationMessage>(OnChangeConfiguration);
+        SubscribeLocalEvent<DoorElectronicsComponent, DoorElectronicsRefreshUiMessage>(OnRefreshUi);
+        SubscribeLocalEvent<DoorElectronicsComponent, InteractUsingEvent>(OnInteractUsing);
+    }
+
+    public void UpdateUserInterface(EntityUid uid, DoorElectronicsComponent component)
+    {
+        List<string> accesses = new List<string>();
```
Comment (AJCM-git) [inline Content.Server/Doors/Electronics/Systems/DoorElectronicsSystem.cs:26]: very weird initializer, just make it a `var`

**Review by AJCM-git — CHANGES_REQUESTED** [review]
I will check the new control and stuff after these

**Comment by Slava0135** [comment]
do we need this after #18638 merge?

**Comment by Errant-4** [comment]
> do we need this after #18638 merge?

That's a rare item, it should be possible to fix/tamper with door permissions without it

File: Resources/Prototypes/Entities/Structures/Doors/Airlocks/base_structureairlocks.yml [hunk 7]
```diff
@@ -48,6 +48,8 @@
   - type: ContainerFill
     containers:
       board: [ DoorElectronics ]
+  - type: AccessReader
+    containerAccessProvider: board
```
[the remark this answers is not in the document]
Comment (c4llv07e) [reply Resources/Prototypes/Entities/Structures/Doors/Airlocks/base_structureairlocks.yml:52]: Done

File: Content.Shared/Doors/Systems/SharedDoorSystem.cs [hunk 8]
```diff
@@ -501,17 +503,17 @@ public bool HasAccess(EntityUid uid, EntityUid? user = null, DoorComponent? door
             TryComp<FirelockComponent>(uid, out var firelock))
             return false;
 
-        if (!Resolve(uid, ref access, false))
+        if (!Resolve(uid, ref access))
             return true;
 
-        var isExternal = access.AccessLists.Any(list => list.Contains("External"));
+        var isExternal = access?.AccessLists.Any(list => list.Contains("External")) ?? false;
```
[the remark this answers is not in the document]
Comment (c4llv07e) [reply Content.Shared/Doors/Systems/SharedDoorSystem.cs]: No, forgot to remove.

File: Content.Shared/Doors/Electronics/DoorElectronicsComponent.cs [hunk 9]
```diff
@@ -0,0 +1,41 @@
+using Robust.Shared.GameStates;
+using Robust.Shared.Serialization;
+
+namespace Content.Shared.Doors.Electronics;
+
+[RegisterComponent]
+public class DoorElectronicsComponent : Component
+{
+    [DataField("accessLevels")]
+    public List<string> AccessLevels = new()
```
[the remark this answers is not in the document]
Comment (c4llv07e) [reply Content.Shared/Doors/Electronics/DoorElectronicsComponent.cs]: Something like this?
```C#
accessLevels = _prototypeManager
            .EnumeratePrototypes<AccessLevelPrototype>()
            .Where(x => x.Name != null)
            .Select(x => x.ID)
            .ToList();
```

File: Content.Server/Doors/Electronics/Systems/DoorElectronicsSystem.cs (hunk 3 above)
Comment (c4llv07e) [reply Content.Server/Doors/Electronics/Systems/DoorElectronicsSystem.cs:84]: I tried, but I get a serialization error and I don't know how to fix it (or what it even means).
`[ERRO] proto: Reading Robust.Shared.Prototypes.EntityPrototype(DoorElectronicsArmory) threw the following exception: Robust.Shared.Serialization.Manager.Exceptions.RequiredFieldNotMappedException: Required field key of type System.String wasn't mapped.`

File: Content.Shared/Doors/Electronics/DoorElectronicsComponent.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (AJCM-git) [reply Content.Shared/Doors/Electronics/DoorElectronicsComponent.cs]: something like that but dont use LINQ

File: Content.Server/Doors/Electronics/Systems/DoorElectronicsSystem.cs (hunk 3 above)
Comment (AJCM-git) [reply Content.Server/Doors/Electronics/Systems/DoorElectronicsSystem.cs:84]: push the code with that error and i will investigate

Comment (c4llv07e) [reply Content.Server/Doors/Electronics/Systems/DoorElectronicsSystem.cs:84]: Here, it's done.

**Review by AJCM-git — CHANGES_REQUESTED** [review]
Sorry for taking so long, was kinda busy but here is why you have prototype errors
image
image

**Comment by c4llv07e** [comment]
Oh, it's weird, but okay. Thank you!
I think that's all for now.

File: Content.Shared/Doors/Electronics/DoorElectronicsEvent.cs [hunk 10]
```diff
@@ -0,0 +1,3 @@
+using Robust.Shared.Serialization;
+
+namespace Content.Shared.Doors.Electronics;
```
Comment (ElectroJr) [inline Content.Shared/Doors/Electronics/DoorElectronicsEvent.cs:3]: Empty file

File: Content.Server/Doors/Electronics/Systems/DoorElectronicsSystem.cs [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+    private void OnChangeConfiguration(
+        EntityUid uid,
+        DoorElectronicsComponent component,
+        DoorElectronicsUpdateConfigurationMessage args)
+    {
+        if (!TryComp<AccessReaderComponent>(uid, out var accessReader))
+        {
+            accessReader = EntityManager.AddComponent<AccessReaderComponent>(uid);
+        }
```
Comment (ElectroJr) [inline Content.Server/Doors/Electronics/Systems/DoorElectronicsSystem.cs:51]: `EnsureComp<AccessReaderComponent>(uid)`

File: Content.Server/Doors/Electronics/Systems/DoorElectronicsSystem.cs [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+        var accesses = new List<string>();
+
+        if (TryComp<AccessReaderComponent>(uid, out var accessReader))
+        {
+            foreach (var accessList in accessReader.AccessLists)
+            {
+                var access = accessList.FirstOrDefault();
+                if (access == null)
+                    continue;
+                accesses.Add(access);
```
Comment (ElectroJr) [inline Content.Server/Doors/Electronics/Systems/DoorElectronicsSystem.cs:34]: I dislike this because this is only half implementing proper access configurations, but fixing this requires improving the UI, which needs doing anyway. So I guess that can just be done in some separate PR.

File: Content.Server/Doors/Electronics/Systems/DoorElectronicsSystem.cs [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+        if (!TryComp<AccessReaderComponent>(uid, out var accessReader))
+        {
+            accessReader = EntityManager.AddComponent<AccessReaderComponent>(uid);
+        }
+
+        accessReader.AccessLists.Clear();
+        foreach (var access in args.accessList)
+        {
+            accessReader.AccessLists.Add(new HashSet<string>(){access});
+        }
```
Comment (ElectroJr) [inline Content.Server/Doors/Electronics/Systems/DoorElectronicsSystem.cs:57]: This should go through some access reader system method instead of modifying it directly.
That method should dirty the component and raise some access changed event.
Door electronics should listen for those events and update the UI.

File: Content.Client/Doors/Electronics/DoorElectronicsBoundUserInterface.cs [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+            }
+        }
+
+        accessLevels.Sort();
+
+        _window = new DoorElectronicsConfigurationMenu(this, accessLevels, _prototypeManager);
+        _window.OnClose += Close;
+        _window.OpenCentered();
+
+        SendMessage(new DoorElectronicsRefreshUiMessage());
```
Comment (ElectroJr) [inline Content.Client/Doors/Electronics/DoorElectronicsBoundUserInterface.cs:38]: Don't send a message, just have the server update the UI state on opening if its required.

**Review by ElectroJr — CHANGES_REQUESTED** [review]

File: Content.Server/Doors/Electronics/Systems/DoorElectronicsSystem.cs (hunk 13 above)
Comment (c4llv07e) [reply Content.Server/Doors/Electronics/Systems/DoorElectronicsSystem.cs]: Something like this?

File: Content.Client/Access/UI/AccessLevelControl.xaml.cs [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+using Robust.Client.UserInterface.Controls;
+using Robust.Client.UserInterface.XAML;
+using Robust.Shared.Prototypes;
+using Content.Shared.Access;
+using Content.Shared.Access.Systems;
+
+namespace Content.Client.Access.UI;
+
+[GenerateTypedNameReferences]
+public sealed partial class AccessLevelControl : GridContainer
```
Comment (AJCM-git) [inline Content.Client/Access/UI/AccessLevelControl.xaml.cs:13]: wont require you to, but can you quickly make the door configurator or whatever that other PR added use this instead of copypasting?, that got past reviews

File: Content.Client/Access/UI/AccessLevelControl.xaml.cs [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+{
+    [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+
+    public readonly Dictionary<string, Button> ButtonsList = new();
+
+    public AccessLevelControl(List<string> accessLevels, IPrototypeManager prototypeManager)
```
Comment (AJCM-git) [inline Content.Client/Access/UI/AccessLevelControl.xaml.cs:19]: update this list to the new prototype id API, it should be something like `List<ProtoId<AccessLevelPrototype>>`, (then update the tryindex)

File: Content.Client/Access/UI/AccessLevelControl.xaml.cs [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+using Content.Shared.Access.Systems;
+
+namespace Content.Client.Access.UI;
+
+[GenerateTypedNameReferences]
+public sealed partial class AccessLevelControl : GridContainer
+{
+    [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+
+    public readonly Dictionary<string, Button> ButtonsList = new();
```
Comment (AJCM-git) [inline Content.Client/Access/UI/AccessLevelControl.xaml.cs:17]: shouldnt this be a new control/boxcontainer in the xaml?

File: Content.Client/Access/UI/AccessLevelControl.xaml.cs [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+                continue;
+            }
+
+            var newButton = new Button
+            {
+                Text = GetAccessLevelName(accessLevel),
+                ToggleMode = true,
+            };
+            AddChild(newButton);
+            ButtonsList.Add(accessLevel.ID, newButton);
```
Comment (AJCM-git) [inline Content.Client/Access/UI/AccessLevelControl.xaml.cs:36]: yeah, looks like it should just add once to a boxcontainer or something here

File: Content.Client/Access/UI/AccessLevelControl.xaml.cs [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    private static string GetAccessLevelName(AccessLevelPrototype prototype)
+    {
+        if (prototype.Name is { } name)
+            return Loc.GetString(name);
+
+        return prototype.ID;
+    }
+    public void UpdateState(List<String> pressedList, List<String>? enabledList = null)
```
Comment (AJCM-git) [inline Content.Client/Access/UI/AccessLevelControl.xaml.cs:47]: newline, also use proper-cased `string`

File: Content.Client/Access/UI/AccessLevelControl.xaml.cs [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+using Robust.Shared.Prototypes;
+using Content.Shared.Access;
+using Content.Shared.Access.Systems;
+
+namespace Content.Client.Access.UI;
+
+[GenerateTypedNameReferences]
+public sealed partial class AccessLevelControl : GridContainer
+{
+    [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
```
Comment (AJCM-git) [inline Content.Client/Access/UI/AccessLevelControl.xaml.cs:15]: you are not injecting anything so this attribute is unnecesary

File: Content.Client/Doors/Electronics/DoorElectronicsBoundUserInterface.cs [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+    private DoorElectronicsConfigurationMenu? _window;
+
+    public DoorElectronicsBoundUserInterface(EntityUid owner, Enum uiKey) : base(owner, uiKey)
+    {
+    }
+
+    protected override void Open()
+    {
+        base.Open();
+        List<string> accessLevels = new();
```
Comment (AJCM-git) [inline Content.Client/Doors/Electronics/DoorElectronicsBoundUserInterface.cs:21]: update

File: Content.Client/Doors/Electronics/DoorElectronicsConfigurationMenu.xaml.cs [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+            button.OnPressed += _ => UpdateConfiguration();
+        }
+    }
+
+    public void UpdateState(DoorElectronicsConfigurationState state)
+    {
+        _buttonsList.UpdateState(state.accessList);
+    }
+
+    private void UpdateConfiguration()
```
Comment (AJCM-git) [inline Content.Client/Doors/Electronics/DoorElectronicsConfigurationMenu.xaml.cs:41]: this is used once so it isnt necessary to make a method, just use _owner.UpdateConfiguration above instead

File: Content.Server/UserInterface/ActivatableUISystem.cs [hunk 23]
```diff
@@ -85,17 +86,33 @@ private void AddOpenUiVerb(EntityUid uid, ActivatableUIComponent component, GetV
...
 
+    private void OnInteractUsing(EntityUid uid, ActivatableUIComponent component, InteractUsingEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        if (component.AllowedItems == null)
+            return;
+
+        if (component.AllowedItems.IsValid(args.Used, EntityManager))
```
Comment (AJCM-git) [inline Content.Server/UserInterface/ActivatableUISystem.cs:110]: wouldnt it be better to move these checks to InteractUI and following the same pattern in here as the ones in OnUseInHand and OnActivate?

File: Resources/Prototypes/Entities/Objects/Devices/Electronics/door_access.yml
Comment (AJCM-git) [inline Resources/Prototypes/Entities/Objects/Devices/Electronics/door_access.yml:1]: this whole file has an unnecesary extra newline between blocks

**Review by AJCM-git — CHANGES_REQUESTED** [review]
Looking good, still needs some cleanup but lets start with the basics

File: Content.Server/UserInterface/ActivatableUISystem.cs (hunk 23 above)
Comment (c4llv07e) [reply Content.Server/UserInterface/ActivatableUISystem.cs:110]: There is a problem with this: `InteractUI` can't check what item was used, and if can't even check whether it was called by item or by hand. 
So, we either add bool `UseOnlyByItems` and pass this item to the `InteractUI` or keep it as it is. I can't say which is better.

File: Content.Client/Access/UI/AccessLevelControl.xaml.cs (hunk 16 above)
Comment (c4llv07e) [reply Content.Client/Access/UI/AccessLevelControl.xaml.cs]: I tried to do this, but as far as I could tell, it would require rewriting the access system api. I can do it, but breaking api is bad, so I don't know if it's worth doing.

**Comment by metalgearsloth** [comment]
Re-open if you come back to this.

**Comment by c4llv07e** [comment]
So, can we merge it now? I think it's ready.

**Comment by metalgearsloth** [comment]
> So, can we merge it now? I think it's ready.

I haven't read it but looks like they're still reviews outstanding.

File: Content.Client/Access/UI/IdCardConsoleBoundUserInterface.cs [hunk 24]
```diff
@@ -23,16 +24,15 @@ public IdCardConsoleBoundUserInterface(EntityUid owner, Enum uiKey) : base(owner
         protected override void Open()
         {
             base.Open();
-            List<ProtoId<AccessLevelPrototype>> accessLevels;
+            List<string> accessLevels;
```
Comment (ShadowCommander) [inline Content.Client/Access/UI/IdCardConsoleBoundUserInterface.cs:27]: Seems like this is a revert. Could you adapt the other code to use `List<ProtoId<AccessLevelPrototype>>` instead of reverting?

File: Content.Client/Access/UI/AccessLevelControl.xaml.cs [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+            }
+
+            var newButton = new Button
+            {
+                Text = GetAccessLevelName(accessLevel),
+                ToggleMode = true,
+            };
+            AddChild(newButton);
+            ButtonsList.Add(accessLevel.ID, newButton);
+        }
```
Comment (ShadowCommander) [inline Content.Client/Access/UI/AccessLevelControl.xaml.cs:35]: Put this code in a new method so that it can be populated after instantiation. Usually other controls call it `Populate`.

This should allow the AccessLevelControl to be defined inside XAML and the Populate method the be called in the constructor. This makes it so that the AccessLevelControl doesn't need to be created in the constructor.

File: Content.Client/Access/UI/AccessLevelControl.xaml.cs (hunk 16 above)
Comment (ShadowCommander) [reply Content.Client/Access/UI/AccessLevelControl.xaml.cs:17]: Seems to have been updated since then. Can you convert them all now?

**Review by ShadowCommander — CHANGES_REQUESTED** [review]

File: Content.Client/Access/UI/IdCardConsoleBoundUserInterface.cs (hunk 24 above)
Comment (c4llv07e) [reply Content.Client/Access/UI/IdCardConsoleBoundUserInterface.cs]: There wasn't `ProtoId` when I created this PR. Maybe I somehow reverted it on merge.
But it doesn't matter, it's all ProtoId now.

File: Content.Client/Access/UI/AccessLevelControl.xaml.cs (hunk 25 above)
Comment (c4llv07e) [reply Content.Client/Access/UI/AccessLevelControl.xaml.cs:47]: Oh, didn't know that. It's very convenient, thank you!

**Comment by c4llv07e** [comment]
Done

**Review by metalgearsloth — COMMENTED** [review]
- You reverted a bunch of arrays to lists when arrays are better so I changed these back.
- Snowflaking in InteractUsing to ActivatableUI is pretty bad but I just want this merged at this point.
